docs/design/hooks.md

```diff
@@ -0,0 +1,322 @@
+        - [Extending Hook](#extending-hook)
+- [Upgrade](#upgrade)
+
+## Summary
+This design is to implement hooks in nfs-provisioner to set annotation or finalizer on NFS PV resources during volume provisioning/deleting events. This document covers the definition of hooks and how to implement this.
+
+## Goals
+Set custom Annotation, Finalizer on NFS resources
+
+## Proposal
+### User Stories
+#### Add custom annotation on NFS resources
+I should be able to configure nfs-provisioner to set the provided annotation on resources created by nfs-provisioner.
+
+#### Add custom finalizer on NFS resources
+I should be able to configure nfs-provisioner to set the provided finalizer on resources created by nfs-provisioner. These Finalizers should exist on resources if not marked to remove on volume deletion event.
+
+#### Add custom annotation and finalizer on NFS resources
+I should be able to configure nfs-provisioner to set the provided annotation and finalizer on resources created by nfs-provisioner.
+
+#### Add custom annotation and finalizer on specific NFS resources
+I should be able to configure nfs-provisioner to set the provided annotation and finalizer on specific resources created by nfs-provisioner.
+
+### Proposed Implementation
+NFS-Provisioner will use user-provided Configmap to learn hooks configuration. These hooks will be executed on volume provisioning or deleting events to set the given information on provided resources. 
+
+### High-Level Design
+User will deploy nfs-provisioner with Configmap having information about hook information. This Configmap should exist in the same namespace in which nfs-provisioner is deployed. Sample nfs-provisioner deployment config is as below:
+
+```yaml
+    spec:
+      serviceAccountName: openebs-maya-operator
+      containers:
+      - name: openebs-provisioner-nfs
+        image: openebs/provisioner-nfs:ci
+        env:
+        - name: OPENEBS_IO_NFS_HOOK_CONFIGMAP
+          value: "nfs-hook"
+        livenessProbe:
+          exec:
+            command:
+            - sh
+            - -c
+            - test `pgrep "^provisioner-nfs.*"` = 1
+          initialDelaySeconds: 30
+          periodSeconds: 60
+```
+
+User needs to provide Configmap name as a value of `OPENEBS_IO_NFS_HOOK_CONFIGMAP` environment variable.
+
+### Low-Level Design
+#### Hook Definition
+NFS Provisioner uses `Provisioner` to provision NFS PV. The existing `Provisioner` definition needs to be extended to use the following hooks definition.
+
+```go
+// ActionType defines type of action for the hook entry
+type ActionType string
+
+const (
+	// ActionAddOnCreateVolumeEvent represent add action on volume create Event
+	ActionAddOnCreateVolumeEvent ActionType = "addOrUpdateEntriesOnCreateVolumeEvent"
+
+	// ActionRemoveOnCreateVolumeEvent represent remove action on volume create Event
+	ActionRemoveOnCreateVolumeEvent ActionType = "removeEntriesOnCreateVolumeEvent"
```

Comment:
Should it be something like below?
```suggestion
	ActionRemoveOnCreateVolumeEvent ActionType = "removeOrUpdateEntriesOnCreateVolumeEvent"
```

Reply:
no. This is a different action.

Reply:
So in future it may get added right?

Reply:
It doesn't make sense to perform remove and updating annotation at same time so this comment doesn't make sense